urlfc1c29991b_nil_mish_InJava_tgz-pJ8-edu_usc_cs_nsl_indexing_test_TestTrieProtoJ8/src/edu/usc/cs/nsl/counting/CountingUtil.java

```diff
@@ -0,0 +1,494 @@
+		for(int i = 0; i < size_range.length; i++){
+			try {
+				bw[i] = new BufferedWriter(new FileWriter(getTmpFileName(size_range[i])));
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		}
+		BufferedReader br = null;
+		try {
+			br = new BufferedReader(new FileReader(inputFile));
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+			System.exit(1);
+		}
+		String the_line = null;
+		try {
+			the_line = br.readLine();
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+		while(the_line != null){
+			String []tokens = the_line.split("\\b\\[A-Z0-9a-z]+\\b");
+			//String []tokens = the_line.split("\\b\\w+\\b");
+			for(String each_token:tokens){
+				int the_length = each_token.length();
+				boolean flag = false;
+				for(int i =0; i < size_range[i]; i++){
+					if(the_length <= size_range[i]){
+						size_count[i] += 1;
+						bw[i].write(each_token+"\n");
+						flag = true;
+						break;
+					}
+				}
+				if(!flag){
+					size_count[size_range.length - 1] += 1;
+					bw[size_range.length - 1].write(each_token+"\n");
+				}
+			}
+			try {
+				the_line = br.readLine();
+			} catch (IOException e) {
+				e.printStackTrace();
+			}
+		}
+		br.close();
+		for(BufferedWriter each_bw: bw){
+			each_bw.close();
+		}
+	}
+	
+	/**
+	 * In order to have good hashing we need a prime number base.
+	 * This function tries to find a close enough approximate
+	 * prime number to the passed value by reducing the range by 1 % 
+	 * each time of the original value till a prime number less
+	 * than the passed value is provided.
+	 * @param the_value The base of the hash function.
+	 * @return the closest prime.
+	 */
+	
+	private int closestPrime(int the_value){
+		float depreciation_factor = 0.01f;
+		int current_value = (int)(the_value * (1.0f-depreciation_factor));
+		while(true){
+			BigInteger bi = BigInteger.valueOf((long) current_value);
+			int result = bi.nextProbablePrime().intValue();
+			if(result <= the_value)
+				return result;
+			current_value = (int)(current_value * (1-depreciation_factor));
+			System.out.println("Stuck here? "+current_value+"\t"+the_value);
+		}
+	}
+
+	/**
+	 * Function to count the words for a given maximum size string in RAM.
+	 * It breaks the count array into chunks. Each chunk size takes into 
+	 * consideration the maximum ram available.
+	 * @param max_string_size maximum string size.
+	 * @return
+	 * @throws IOException
+	 */
+
+	public int countInRAM(int max_string_size) throws IOException{
+		int num_of_chunks = 0;
+		int chunk_size = closestPrime(max_ram/(max_string_size + 8));
+		
+		/* We need to read in long as the file may only have a 
+		 * single alphabet repeated multiple times which will then 
+		 * exceed the INTEGER.MAX_VALUE limit assuming we have a 32 bit JVM.
+		 */		
+		long[] the_big_array = new long[big_array_size];   		
+		BufferedReader br = null;
+		br = new BufferedReader(new FileReader(getTmpFileName(max_string_size)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *br*. The resource is referenced at lines 192, 193, 199. The resource is closed at line 202. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *br* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).